hw11/hw11_t02.py

```diff
@@ -0,0 +1,32 @@
+"""
+You are given the following code.
+
+class Order:
+    morning_discount = 0.25
+    def __init__(self, price):
+        self.price = price
+    def final_price(self):
+        return self.price - self.price * self.morning_discount
+Make it possible to use different discount programs.
+Hint: use strategy behavioural OOP pattern.
+https://refactoring.guru/design-patterns/strategy
+Example of the result call:
+def morning_discount(order):
+    ...
+def elder_discount(order):
+    ...
+order_1 = Order(100, morning_discount)
+assert order_1.final_price() == 50
+order_2 = Order(100, elder_discount)
+assert order_1.final_price() == 10
+"""
+from typing import Callable
+
+
+class Order:
+    def __init__(self, price: int, discount_program: Callable):
+        self.price = price
+        self.discount_program = discount_program()
```

Comment:
This looks way too complex, instead of passing the function and call it immediately you can just pass the result of the function call and the code will be more simple.

Reply:
> you can just pass the result of the function call and the code will be more simple.

And now I completely don't understand the meaning of this task and strategies.
```
def morning_discount(order):
    ...
```
Why `order` is here? What should I change? Discount programs or `Order` class?

Reply:
Strategy pattern means that you outsource some part of the work.  When you asked to calculate the order price, you delegate it to the strategy.   

In your implementation, you just hardcoded the discount logic in your code. `self.discount_program` is just a price multiplier.  

Let's add a test for a discount that removes fixed value.

```python
order_1 = Order(50, minus_one_point)
assert order_1.final_price() == 49.0  
```

Reply:
The idea of the strategy pattern is what you can have all discount mentioned above without changing Order class.
For each type of discount, you just write a new function and pass it to the Order.

I found a good example of a strategy pattern that widely used in Python. https://docs.python.org/3/howto/sorting.html#key-functions You can change sorting behavior without changing sort function itself.

Reply:
Now I see the idea more clearly. I transferred all calculations to incoming functions. Check out my last commit, please.